interfaces/builtin/docker_support.go

```diff
@@ -161,12 +161,167 @@ pivot_root,
 # use 'privileged-containers: true' to support --security-opts
 
 # defaults for docker-default
-change_profile unsafe /** -> docker-default,
+# unfortunately, the docker snap is currently (by design?) setup to have both 
+# the privileged and unprivileged variant of the docker-support interface 
+# connected which means we rules that are compatible to allow both transitioning
+# to docker-default profile here AAAAAAND transitioning to any other profile 
+# below in the privileged snippet, BUUUUUUUT also need to be triply compatible
+# with the injected compatibility snap-confine transition rules to temporarily
+# support executing other snaps from devmode snaps. So we are left with writing
+# out these extremely verbose regexps because AARE does not have a negative 
+# concept to exclude just the paths we want. 
+# When we drop support for executing other snaps from devmode snaps this can go
+# back to the much simpler rule:
+# change_profile unsafe /** -> docker-default,
```

Comment:
Are we sure we need to apply the "monster trick" in this case?

I'm asking because what we have had so far is:
```
change_profile unsafe /** -> docker-default,
[...]
change_profile unsafe /** -> cri-containerd.apparmor.d,
[...]
change_profile unsafe /**,
```
Given that this interface is already privileged, is there any harm in keeping only the last rule (`change_profile unsafe /**,`), which I assume gives the permissions to transition to any profile?

Reply:
So there are a few conflating problems which led me to do this, if we addressed one of the other problems we could probably forgo some of these monster rules

1. The docker-support interface when used with the `privileged-containers: true` attribute includes all of the non-privileged containers rules and the extra privileged snippets, meaning that it is intended to be used standalone (i.e. it was expected you either had privileged connected or the normal unprivileged one, not both at the same time)
2. The docker snap is currently setup to always auto-connect both interfaces, I have a feeling that was not intentional, as the way it was described to me originally was that specific users/customers had the option to run privileged containers by connecting the interface
3. Due to 2, we probably cannot change or remove anything about how it works today as this could break existing users of the docker snap, of which there are many

What would be ideal is any of the following:
* Only connect one of the interfaces at a time (make the flavors mutually exclusive), this way snapd could have code which just specifies the more generic rules since those allow the more specific ones anyways
* Make the privileged variant only add the extra snippet and not duplicate all the rest of the docker-support snippet again (though this doesn't really hurt anything to have duplicated rules, the parser can de-duplicate them)
* have code inside snapd which detects that both the privileged and the unprivileged are connected simultaneously and then omit the specific rules which are allowed by the more generic ones

The last option is the most feasible in terms of being allowed (i.e. not breaking users), however it is also the most tricky since it would require iterating over all the lists of other connected plugs that an app has, which AFAICT would be tricky codewise to find out (since you can't just look at the app's plugs, you need to know the _connected_ plugs).

All of that is to say that I found the easiest way to fix this was just to make sure that all the rules are always compatible with each other no matter the combination of plugs/attributes. That led me down this path of de-conflicting:

1. First I found that the `/** rwlix,` rule conflicted with the snap-confine file glob rules, so I made that one into the big tower
2. Then I realized that the tower rule for `/** rwlix,` conflicted with the `change_profile unsafe /**` rule, so I towerized that one
3. Then I found that the towerized/monster version of `change_profile unsafe /**` conflicts with the specific rules for `change_profile unsafe /** -> docker-default,` (and `cri-containerd.apparmor.d` too), so I towerized that one too

QED. :smiley:

Reply:
I see, thanks for the explanation!

I can see some value in letting an interface know what other plugs/sockets are connected. I'll try to bring this to the forums eventually.